Scripts/Playthrough.cs

```diff
@@ -374,6 +341,208 @@ private bool IsConnected()
+
+            _room.OnMessage<string>("status", (status) =>
+            {
+                Logger.Log($"Received `status` event: {JsonConvert.SerializeObject(status)}");
+                if (status == "ready")
+                {
+                    Logger.Log("Ready to begin play");
+                    OnReady();
+                }
+            });
+
+            _room.OnMessage<Events.MessageEvent>("message", (message) =>
+            {
+                Logger.Log($"Received `message` event: {JsonConvert.SerializeObject(message)}");
+                OnMessage?.Invoke(message);
+            });
+
+            _room.OnMessage<Events.StartTypingEvent>("start-typing", (message) =>
+            {
+                Logger.Log($"Received `start-typing` event: {JsonConvert.SerializeObject(message)}");
+                OnStartTyping?.Invoke(message);
+            });
+
+            _room.OnMessage<Events.StopTypingEvent>("stop-typing", (message) =>
+            {
+                Logger.Log($"Received `stop-typing` event: {JsonConvert.SerializeObject(message)}");
+                OnStopTyping?.Invoke(message);
+            });
+
+            _room.OnMessage<Events.ProblemEvent>("problem", (message) =>
+            {
+                Logger.LogWarning($"Received `problem` event: {JsonConvert.SerializeObject(message)}");
+            });
+
+            _room.OnMessage<string>("pong", (message) =>
+            {
+                _pingCount = 0;
+                OnPingSuccess?.Invoke();
+                SetConnectionState(ConnectionState.Connected);
+            });
+
+            _room.OnLeave += (code) =>
+            {
+                Logger.Log($"Connection closed. Attempting to reconnect to Playthrough.");
+                MainThreadConsumer.Instance?.StartCoroutine(TryToReconnect());
+            };
+        }
+
+        private void OnReady()
+        {
+            SetConnectionState(ConnectionState.Connected);
+
+            _onReadyCallback?.Invoke();
+
+            _calledByDisconnect = false;
+            _reconnectionTryCount = 0;
+
+            Task.Run(FirePing);
+        }
+
+        private IEnumerator TryToReconnect()
+        {
+            Logger.Log("Trying to reconnect.");
+            if (_calledByDisconnect)
+            {
+                SetConnectionState(ConnectionState.Disconnected);
+                yield break;
+            }
+
+            // cannot allow multiple reconnection coroutines to run concurrently
+            if (IsReconnecting())
+            {
+                yield break;
+            }
+
+            SetConnectionState(ConnectionState.Reconnecting);
+
+            MainThreadConsumer.Instance.StartCoroutine(Reconnect());
+        }
+
+        private bool IsReconnecting()
+        {
+            return _connectionState == ConnectionState.Reconnecting;
+        }
+
+        private IEnumerator Reconnect()
+        {
+            ColyseusRoom<object> room = null;
+
+            while (room == null)
```

Comment:
maybe `while(_room == null)` instead rather than creating a new variable `room`?

Reply:
I added this because we're technically still referencing the original `_room` when issuing the Reconnect task via the ID and Session. I can store those IDs locally in separate variables ofc, but seemed more convenient to have this